edgedb-tokio/src/tutorial.md

```diff
@@ -0,0 +1,725 @@
+
+```rust
+#[derive(Debug, Deserialize, Queryable)]
+#[edgedb(json)]
+pub struct JsonQueryableAccount {
+    pub username: String,
+    pub id: Uuid,
+}
+
+let json_queryable_accounts: Vec<JsonQueryableAccount> = client
+    .query("select <json>Account { username, id }", &())
+    .await
+    .unwrap();
+```
+
+This attribute can also be used on an inner struct value that implements Queryable. Here, some random json is turned into a HashMap<String, String>:
+
+```rust
+#[derive(Debug, Deserialize, Queryable)]
+pub struct InnerJsonQueryableAccount {
+    pub username: String,
+    pub id: Uuid,
+    #[edgedb(json)]
+    pub some_json: HashMap<String, String>,
+}
+
+let query = r#" with j := <json>(
+    nice_user := "yes",
+    bad_user := "no"
+)
+select Account {
+    username,
+    id,
+    some_json := j
+    };"#;
+let query_res: Vec<InnerJsonQueryableAccount> = client.query(query, &()).await.unwrap();
+``` 
+
+## Transactions
+
+The client also has a .transaction() method that allows atomic [transactions](https://www.edgedb.com/docs/edgeql/transactions). Wikipedia has a good example of a transaction and why it would be best done atomically:
+
+```
+An example of an atomic transaction is a monetary transfer from bank account A 
+to account B. It consists of two operations, withdrawing the money from account A 
+and saving it to account B. Performing these operations in an atomic transaction 
+ensures that the database remains in a consistent state, that is, money is 
+neither lost nor created if either of those two operations fails.
+```
+
+A transaction removing 10 cents from one customer's account and placing it in the other's account would look as follows:
+
+```rust
+#[derive(Debug, Deserialize, Queryable)]
+#[edgedb(json)]
+pub struct BankCustomer {
+    pub name: String,
+    pub bank_balance: i32,
+}
+    // Transactions
+    // Customer1 has an account with 110 cents in it.
+    // Customer2 has an account with 90 cents in it.
+    // Customer1 is going to send 10 cents to Customer 2. This will be a transaction 
+    // because we don't want the case to ever occur - even for a split second -  
+    // where one account has sent money while the other has not received it yet.
+
+    // After the transaction is over, each customer should have 100 cents.
+
+    let customers_after_transaction = client.transaction(|mut conn| async move {
+        let res_1 = conn.query_required_single_json
+        ("select(update BankCustomer filter .name = <str>$0 set 
+        { bank_balance := .bank_balance - 10 }){name, bank_balance};", &("Customer1",)).await?;
+        let res_2 = conn.query_required_single_json
+        ("select(update BankCustomer filter .name = <str>$0 set
+        { bank_balance := .bank_balance + 10 }){name, bank_balance};", &("Customer2",)).await?;
+        Ok(vec![res_1, res_2])
+    }).await?;
+```
+
+## Links can work instead of transactions
```

Comment:
I think I agree with @tailhook, but I'd love to see this make it to the general documentation. Maybe at https://www.edgedb.com/docs/edgeql/transactions ? If so, maybe it would be good to also include your bank transaction example (generalized so that it's not about Rust) first on that page as a non-link example of a transaction.

Reply:
I think I'll just shrink this to a general comment about how links are often the more idiomatic (and easier) way to handle an action that might seem like a transaction at first. This goes back to the discussion where Paul mentioned a link instead of a transaction that I found helpful so I do want to keep it in as a tip.

And yeah, add something to the general documentation too I think. I'll make a PR for that.

Reply:
Okay, I've cut that part at the end to the bare minimum to just let the readers do their own exploration from there. Will bring it back to life in the general transaction documentation in a similar form.